idx/idx.go

```diff
@@ -1,363 +1,30 @@
-func (i *Idx) QueryRadix(org int, query string, equal bool) ([]Glob, []*schema.MetricDefinition) {
-	var globs []Glob
-	var defs []*schema.MetricDefinition
-	l := len(query)
-	seen := make(map[string]bool)
-	fn := func(k string, v interface{}) bool {
-		// figure out if we're a leaf or not
-		dot := strings.IndexByte(k[l:], '.')
-		var leaf bool
-		m := k
-		if dot == -1 {
-			leaf = true
-		} else {
-			m = k[:dot+l]
-		}
-
-		if len(m) > len(query) && equal {
-			return false
-		}
-		if !seen[m] {
-			seen[m] = true
-			globs = append(globs, Glob{Metric: MetricKey(m), IsLeaf: leaf})
-			defs = append(defs, &i.defs[int(v.(DocID))])
-		}
-		// false == "don't terminate iteration"
-		return false
-	}
-	i.WalkPrefix(-1, query, fn)
-	i.WalkPrefix(org, query, fn)
-	return globs, defs
-}
-
-// output is unsorted, if you want it sorted, do it yourself
-func (i *Idx) Match(org int, query string) (MatchType, []Glob, []*schema.MetricDefinition) {
-
-	// exact match only, could still result in multiple definitions.
-	// the radix tree is a good option for this case
-	if !strings.ContainsAny(query, "*{}[]?") {
-		globs, defs := i.QueryRadix(org, query, true)
-		return MatchLiteral, globs, defs
-	}
-
-	// only one trailing star -> prefix match
-	if strings.Index(query, "*") == len(query)-1 && !strings.ContainsAny(query, "{}[]?") {
-		query = query[:len(query)-1]
-		globs, defs := i.QueryRadix(org, query, false)
-		return MatchPrefix, globs, defs
-	}
-
-	// at least one interior star or a more complicated query
-	globs, defs := i.QueryTrigrams(org, query)
-	return MatchTrigram, globs, defs
-}
-
-func extractTrigrams(query string) []trigram.T {
-
-	if len(query) < 3 {
-		return nil
-	}
-
-	var start int
-	var i int
-
-	var trigrams []trigram.T
-
-	for i < len(query) {
-		if query[i] == '[' || query[i] == '{' || query[i] == '*' || query[i] == '?' {
-			trigrams = trigram.Extract(query[start:i], trigrams)
-
-			if query[i] == '[' {
-				for i < len(query) && query[i] != ']' {
-					i++
-				}
-			}
-
-			if query[i] == '{' {
-				for i < len(query) && query[i] != '}' {
-					i++
-				}
-			}
-
-			start = i + 1
-		}
-		i++
-	}
-
-	trigrams = trigram.Extract(query[start:i], trigrams)
-
-	return trigrams
+type Node struct {
+	Path string
+	Leaf bool
+	Defs []schema.MetricDefinition
 }
 
-// Prune is a custom pruning function for the trigram index
-// it removes all trigrams that are present in more than the specified percentage of the documents.
-// However, it makes sure that neither tAllDocIDs,
-// nor any of the org specific trigrams used for access control are ever pruned.
-func (i *Idx) Prune(pct float64) {
-	maxDocs := int(pct * float64(len(i.tridx[trigram.TAllDocIDs])))
-	for k, v := range i.tridx {
-		// all values >= 1<<31 are special trigrams that should be kept
-		if uint32(k) < uint32(2147483648) && len(v) > maxDocs {
-			i.tridx[k] = nil
-		}
-	}
+type MetricIndex interface {
+	Init(met.Backend) error
+	Stop()
+	Add(*schema.MetricData)
```

Comment:
we should probably document explicitly that it's safe for the caller to change the pointed-to data, after the Add call returns.